core/src/main/java/org/elasticsearch/index/search/NestedHelper.java

```diff
@@ -0,0 +1,161 @@
+ */
+
+package org.elasticsearch.index.search;
+
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.BoostQuery;
+import org.apache.lucene.search.ConstantScoreQuery;
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.MatchNoDocsQuery;
+import org.apache.lucene.search.PointRangeQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.TermQuery;
+import org.apache.lucene.search.BooleanClause.Occur;
+import org.elasticsearch.index.mapper.MapperService;
+import org.elasticsearch.index.mapper.ObjectMapper;
+
+public final class NestedHelper {
+
+    private final MapperService mapperService;
+
+    public NestedHelper(MapperService mapperService) {
+        this.mapperService = mapperService;
+    }
+
+    /** Returns true if the given query might match nested documents. */
+    public boolean mightMatchNestedDocs(Query query) {
+        if (query instanceof ConstantScoreQuery) {
+            return mightMatchNestedDocs(((ConstantScoreQuery) query).getQuery());
+        } else if (query instanceof BoostQuery) {
+            return mightMatchNestedDocs(((BoostQuery) query).getQuery());
+        } else if (query instanceof MatchAllDocsQuery) {
+            return true;
+        } else if (query instanceof MatchNoDocsQuery) {
+            return false;
+        } else if (query instanceof TermQuery) {
+            // We only handle term queries and range queries, which should already
+            // cover a high majority of use-cases
+            return mightMatchNestedDocs(((TermQuery) query).getTerm().field());
+        } else if (query instanceof PointRangeQuery) {
+            return mightMatchNestedDocs(((PointRangeQuery) query).getField());
+        } else if (query instanceof BooleanQuery) {
+            BooleanQuery bq = (BooleanQuery) query;            
+            boolean hasRequiredClauses = bq.clauses().stream().anyMatch(BooleanClause::isRequired);
+            if (hasRequiredClauses) {
+                return bq.clauses().stream()
+                        .filter(BooleanClause::isRequired)
+                        .map(BooleanClause::getQuery)
+                        .allMatch(this::mightMatchNestedDocs);
+            } else {
+                return bq.clauses().stream()
+                        .filter(c -> c.getOccur() == Occur.SHOULD)
+                        .map(BooleanClause::getQuery)
+                        .anyMatch(this::mightMatchNestedDocs);
+            }
+        } else if (query instanceof ESToParentBlockJoinQuery) {
+            return ((ESToParentBlockJoinQuery) query).getPath() != null;
+        } else {
+            return true;
+        }
+    }
+
+    /** Returns true if a query on the given field might match nested documents. */
+    boolean mightMatchNestedDocs(String field) {
+        if (field.startsWith("_")) {
+            // meta field
+            return true;
```

Comment:
I did not implement the logic since every meta field can have its own logic when it comes to meta fields, eg. _uid is the same for nested docs, _type is used for nested paths, and other fields are not added to nested docs but maybe some of them will one day, eg. _field_names. So I did not implement it for now and returned `true`, which is always safe to do (we might just apply filters when they are not needed). I agree I should have at least put a comment about it however.